patches/tModLoader/Terraria/GameContent/TeleportPylonsSystem.cs.patch

```diff
@@ -1,11 +1,209 @@
+ using Terraria.Chat;
+ using Terraria.DataStructures;
+ using Terraria.GameContent.NetModules;
+ using Terraria.GameContent.Tile_Entities;
++using Terraria.ID;
+ using Terraria.Localization;
++using Terraria.ModLoader;
+ using Terraria.Net;
+ 
+ namespace Terraria.GameContent
+@@ -14,11 +_,15 @@
+ 	{
+ 		private List<TeleportPylonInfo> _pylons = new List<TeleportPylonInfo>();
  		private List<TeleportPylonInfo> _pylonsOld = new List<TeleportPylonInfo>();
++		private List<TeleportPylonInfo> _vanillaPylons = new List<TeleportPylonInfo>();
++		private List<TeleportPylonInfo> _moddedPylons = new List<TeleportPylonInfo>();
  		private int _cooldownForUpdatingPylonsList;
  		private const int CooldownTimePerPylonsListUpdate = int.MaxValue;
 -		private SceneMetrics _sceneMetrics = new SceneMetrics();
 +		internal SceneMetrics _sceneMetrics = new SceneMetrics();
  
- 		public List<TeleportPylonInfo> Pylons => _pylons;
+-		public List<TeleportPylonInfo> Pylons => _pylons;
++		public List<TeleportPylonInfo> AllPylons => _pylons;
++		public List<TeleportPylonInfo> VanillaPylons => _vanillaPylons;
++		public List<TeleportPylonInfo> ModdedPylons => _moddedPylons;
+ 
+ 		public void Update() {
+ 			if (Main.netMode != 1) {
+@@ -32,7 +_,12 @@
+ 			}
+ 		}
+ 
+-		public bool HasPylonOfType(TeleportPylonType pylonType) => _pylons.Any((TeleportPylonInfo x) => x.TypeOfPylon == pylonType);
++		public bool HasPylonOfType(TeleportPylonType pylonType, int tileType = TileID.TeleportationPylon) {
++			if (pylonType == TeleportPylonType.Modded && tileType >= TileID.Count) {
++				return _pylons.Any(x => !string.IsNullOrEmpty(x.ModName) && ModContent.Find<ModPylon>(x.ModName, x.ModPylonName).Type == tileType);
++			}
++			return _pylons.Any(x => x.TypeOfPylon == pylonType);
++		}
+ 		public bool HasAnyPylon() => _pylons.Count > 0;
+ 
+ 		public void RequestImmediateUpdate() {
+@@ -42,17 +_,33 @@
+ 			}
+ 		}
+ 
++		private void AddNewPylon(TeleportPylonInfo info) {
++			if (info.TypeOfPylon == TeleportPylonType.Modded)
++				_moddedPylons.Add(info);
++			else
++				_vanillaPylons.Add(info);
++			_pylons.Add(info);
++		}
++
+ 		private void UpdatePylonsListAndBroadcastChanges() {
+ 			Utils.Swap(ref _pylons, ref _pylonsOld);
+ 			_pylons.Clear();
++			_vanillaPylons.Clear();
++			_moddedPylons.Clear();
+ 			foreach (TileEntity value in TileEntity.ByPosition.Values) {
+-				TETeleportationPylon tETeleportationPylon = value as TETeleportationPylon;
+-				if (tETeleportationPylon != null && tETeleportationPylon.TryGetPylonType(out TeleportPylonType pylonType)) {
+-					TeleportPylonInfo teleportPylonInfo = default(TeleportPylonInfo);
+-					teleportPylonInfo.PositionInTiles = tETeleportationPylon.Position;
++				TeleportPylonType pylonType = TeleportPylonType.Modded;
++				if (value.GetType().IsDefined(typeof(IsPylonEntityAttribute), true) || (value is TETeleportationPylon tETeleportationPylon && tETeleportationPylon.TryGetPylonType(out pylonType)))  {
```

Comment:
`IPylonTileEntity` marker interface rather than attribute please. Add it to vanilla's TE too. There's also really no reason for `TryGetPylonTypeFromTileCoords` in `TETeleportationPylon` not be `static`. And you could get the pylon type from the modded tile there.